client/starwhale/core/dataset/store.py

```diff
@@ -438,19 +453,151 @@ def sign_uri(self, uri: str) -> str:
         return r["data"].get(uri, "")  # type: ignore
 
 
-class S3BufferedFileLike:
+_BFType = t.TypeVar("_BFType", bound="BaseBufferedFileLike")
+
+
+class BaseBufferedFileLike(metaclass=ABCMeta):
+    def __init__(self, buffer_size: int) -> None:
+        self._read_buffer = BytesBuffer(buffer_size)
+
+    def __enter__(self: _BFType) -> _BFType:
+        return self
+
+    def __exit__(
+        self,
+        type: t.Optional[t.Type[BaseException]],
+        value: t.Optional[BaseException],
+        trace: TracebackType,
+    ) -> None:
+        if value:
+            print(f"type:{type}, exception:{value}, traceback:{trace}")
+
+        self.close()
+
+    def read(self, size: int) -> bytes:
+        ret = self._read(size)
+        if isinstance(ret, memoryview):
+            return ret.tobytes()
+        elif isinstance(ret, bytes):
+            return ret
+        else:
+            raise FormatError(f"{ret}:{type(ret)} is not bytes or memoryview type")
+
+    def _read(self, size: int) -> t.Union[memoryview, bytes]:
+        if size == 0:
+            return b""
+        elif size < 0:
+            return self._read_buffer.read() + self._read_exhausted_raw()
+        else:
+            while len(self._read_buffer) < size:
+                fill_size = self._fill_new_bytes()
+                if fill_size == 0:
+                    return self._read_buffer.read()
+
+            return self._read_buffer.read(size)
+
+    @abstractmethod
+    def _read_exhausted_raw(self) -> bytes:
+        raise NotImplementedError
+
+    @abstractmethod
+    def _fill_new_bytes(self) -> int:
+        raise NotImplementedError
+
+    def close(self) -> None:
+        self._read_buffer.close()
+
+
+class HttpBufferedFileLike(BaseBufferedFileLike):
+    def __init__(
+        self,
+        url: str,
+        headers: t.Optional[t.Dict[str, str]] = None,
+        timeout: int = 90,
+        buffer_size: int = 256 * 1024,
+    ) -> None:
+        super().__init__(buffer_size)
+        if headers is None:
+            self.headers = {"Accept-Encoding": "identity"}
+        else:
+            self.headers = headers
+
+        self.buffer_size = buffer_size
+        self.resp = requests.get(
+            url,
+            stream=True,
+            headers=self.headers,
+            timeout=timeout,
+        )
+
+        if not self.resp.ok:
+            self.resp.raise_for_status()
+
+        self._read_iter = self.resp.iter_content(buffer_size)
+
+    def _read_exhausted_raw(self) -> bytes:
+        return self.resp.raw.read()  # type: ignore
```

Comment:
Will it be better that `HttpBufferedFileLike` overrides the `read` function and just return self.resp.raw.read(size)?

Reply:
one read case:

```
read(32)  <-- first read 32 bytes
read(8)    <-- then read 8 bytes
read(-1)   <--  last read all remaining bytes
```
from the bellowing example, some bytes already maybe read into the buffer before , so we cannot read `self.resp.raw.read` directly.

Another we use the buffer for http stream read.

Reply:
updated